### ISSUE TITLE
ogr2ogr -append silently upgrades field length instead of truncating values

### ISSUE THREAD

**Issue by uclaros** [OSGeo/gdal#3709]
## Expected behavior and actual behavior.
When using `ogr2ogr -append` to append features to an empty shapefile, if the source string fields are wider than the target fields then I expect some strings to be truncated and the target schema to be preserved.
The actual behavior is that the target string fields are widened in order to accommodate the larger source strings

## Steps to reproduce the problem.
- create a shapefile `tgt.shp` with a string field `str` , length 5
- create a shapefile `src.shp` with a string field `str` , length 7
- add a feature to `src.shp`, set `str` to `1234567`
- ogr2ogr -append tgt.shp src.shp src
- ogrinfo tgt.shp
- field `str` has now length 7

## Operating system
Debian Bullseye

## GDAL version and provenance
GDAL 3.2.1, released 2020/12/29 , from debian testing repo

**Comment by rouault** [comment]
The current behavior is intended. People would complain of the opposite of yours otherwise :-)

**Comment by uclaros** [comment]
> People would complain of the opposite of yours otherwise :-)

They would be wrong to complain though... wouldn't they?
You wouldn't even think of doing this if the target was on Postgres!
Silently changing the schema definition is *not nice*... It actually defeats the purpose of having a field length!

**Comment by rouault** [comment]
In shapefile, field width is more an artificial limit due to how the format is structured. In particular, when GDAL creates a new shapefile from other datasources that don't specify a field width, it defaults to 80 in a completely arbitrary way. This actual behaviour of growing shapefile column width is actually a behavior of the writer side of shapefile driver itself (ogr2ogr isn't aware of it) and was implemented quite a long time ago (8e7f9d2d4f077a34d83ea9066b1a856f10b81ea2). Most people prefer growing than truncation

**Comment by uclaros** [comment]
Thanks for the extra info.
This makes sense for layer creation, hence the `-lco RESIZE=YES` mentioned in your linked PR.
Appending though is different since there is an existing field width and, even if it is artificial, one would expect it to be preserved and the user warned.
Especially after receiving warnings like `Warning 1: Value 1779.56207139252183 of field LEN of feature 43 not successfully written. Possibly due to too larger number with respect to field width`

**Comment by jratike80** [comment]
As a user I would rather expect that fields would be widened also in the PostgreSQL case. As a PostgreSQL admin I would think in the opposite way. Users, as I know them, are mostly interested in getting their data stored.

> They would be wrong to complain though... wouldn't they?

No, I wouldn't say that they are wrong. I would not say that you are wrong either. There are two sides which are both wrong or right depending on how you look at the situation. But we must select one or the other alternative and I would keep the current behavior.

**Comment by uclaros** [comment]
> As a user I would rather expect that fields would be widened also in the PostgreSQL case. As a PostgreSQL admin I would think in the opposite way. Users, as I know them, are mostly interested in getting their data stored.

I totally agree, though none of such users that I've met would ever use `ogr2ogr`, that's why I expect such a console tool to be exact and consistent.

> But we must select one or the other alternative and I would keep the current behavior.

Selecting the one behavior for one format and the other for another format is inconsistent behavior... And unlike right and wrong, consistency has no such room for interpretation.
Also, the warning `Warning 1: Value 1779.56207139252183 of field LEN of feature 43 not successfully written. Possibly due to too larger number with respect to field width` hints that some kind of truncation has taken place for a specific field type. So truncating some types while resizing some others is not the best in terms of consistency.

**Comment by jratike80** [comment]
It is documented in https://gdal.org/drivers/vector/shapefile.html that auto-extend affects only string and integer fields.

> Field sizes 
The driver knows to auto-extend string and integer fields (up to the 255 bytes limit imposed by the DBF format) to dynamically accommodate for the length of the data to be inserted.
>
> It is also possible to force a resize of the fields to the optimal width by issuing a SQL ‘RESIZE <tablename>’ via the datasource ExecuteSQL() method. This is convenient in situations where the default column width (80 characters for a string field) is bigger than necessary.

GDAL does offer quite a lot of abstraction for all the formats that it handles. Being totally consistent with all of them and with any aspect is impossible (you do not want to limit field names into max 10 characters because shapefile has that limit). In this case I honestly think that the current behavior of shapefile driver to widen the field if it can is better for more users than the other way. I probably would not call that alternative as exact because truncation would alter the data while it preserves the schema.

**Comment by uclaros** [comment]
Why not change an int field to string when trying to append strings to it then?

**Comment by rouault** [comment]
A few observations:
- the current behaviour should remain the default one. I claim that column width in .dbf doesn't reflect most of the time an intented semantics of a limitation of size, but is a contingency of the format
- we could *possibly* introduce a shapefile specific open option to ask to not modify the schema (like -doo ALTER_SCHEMA=NO). But would you use that ?

**Comment by uclaros** [comment]
That was actually no 1 on my suggestion list after seeing you actually prefer the current behavior, I would definitely use that!
I would not make it shapefile specific though, it should apply to all formats that alter the schema to fit the data fed.

**Comment by rouault** [comment]
> I would not make it shapefile specific though, it should apply to all formats that alter the schema to fit the data fed.

This is really driver specific behavior. From my memory, I can't name another driver that would do that, but I wouldn't exclude it either. I can't think of a better solution that a dataset open option, with a generic name that could be reused in another driver.
Tagging this as enhancement, but very low priority AFAIC

**Comment by jratike80** [comment]
The PostGIS driver has an option `PRECISION=YES/NO` https://gdal.org/drivers/vector/pg.html. It is YES by default and it keeps the schema.
There are some users and developers who think that the default should be NO  
https://github.com/OSGeo/gdal/issues/3345 and 
https://github.com/npgsql/npgsql/issues/3471


> Why not change an int field to string when trying to append strings to it then?
I believe this is a hyperbole https://en.wikipedia.org/wiki/Hyperbole, right?

**Comment by uclaros** [comment]
> I believe this is a hyperbole

Maybe it is, but its answer can also be used as an argument against the current behavior with string fields!

**Comment by jratike80** [comment]
Then I have expressed myself very badly. Widening the field width does not change the existing data. 123 remains 123 even if the shapefile accepts also a new record with value 1234567. Changing the data type from integer into strings would update the existing records into strings as well and that is not acceptable.

Adding strings into integer field is possible and correct in SQLite but that is off-topic.

**Comment by uclaros** [comment]
Ok, sorry for insisting, I'll just accept the special snowflake shapefile driver!
Anyways, the shapefile driver docs need to be updated:
> The field width and precision are directly used to establish storage size in the .dbf file. This means that strings longer than the field width, or numbers that don’t fit into the indicated field format will suffer truncation.

Should be saying instead that string field width plays no role whatsoever as it is automatically expanded to accommodate longer strings
